thirdparty/cmdconfig/commands/cmdeval/cmdeval.go

```diff
@@ -164,43 +175,58 @@ func (r *EvalFnRunner) NewFunction() *kptfile.Function {
 	return newFn
 }
 
-// SaveFnToKptfile adds the evaluated function and its arguments to Kptfile `pipeline.mutators` list.
-func (r *EvalFnRunner) SaveFnToKptfile() {
+// Add the evaluated function to the kptfile.Function list, this Function can either be
+// `pipeline.mutators` or `pipeline.validators`
+func (r *EvalFnRunner) updateFnList(oldFNs []kptfile.Function) []kptfile.Function {
+	var newFns []kptfile.Function
 	pr := printer.FromContextOrDie(r.Ctx)
-	kf, err := pkg.ReadKptfile(filesys.FileSystemOrOnDisk{}, r.RunFns.Path)
-	if err != nil {
-		pr.Printf("function not added: Kptfile not exists\n")
-		return
-	}
-	// TODO(yuwenma): Right now we cannot tell if a function is a mutator or validator. Once kpt supports
-	// OCI images, we can add annotations to image and find out the function type from these annotations.
-	if kf.Pipeline == nil {
-		kf.Pipeline = &kptfile.Pipeline{}
-	}
-	if kf.Pipeline.Mutators == nil {
-		kf.Pipeline.Mutators = []kptfile.Function{}
-	}
-	var newMutators []kptfile.Function
 	found := false
-	for _, m := range kf.Pipeline.Mutators {
+	for _, m := range oldFNs {
 		switch {
 		case m.Image != "" && m.Image == r.Image:
 			pr.Printf("found image in Kptfile, updating...\n")
-			newMutators = append(newMutators, *r.NewFunction())
+			newFns = append(newFns, *r.NewFunction())
 			found = true
 		case m.Exec != "" && m.Exec == r.Exec:
 			pr.Printf("found exec in Kptfile, updating...\n")
-			newMutators = append(newMutators, *r.NewFunction())
+			newFns = append(newFns, *r.NewFunction())
 			found = true
 		default:
-			newMutators = append(newMutators, m)
+			newFns = append(newFns, m)
 		}
 	}
 	if !found {
-		newMutators = append(newMutators, *r.NewFunction())
+		newFns = append(newFns, *r.NewFunction())
 		pr.Printf("adding function to Kptfile\n")
 	}
-	kf.Pipeline.Mutators = newMutators
+	return newFns
+}
+
+// SaveFnToKptfile adds the evaluated function and its arguments to Kptfile `pipeline.mutators` list.
+func (r *EvalFnRunner) SaveFnToKptfile() {
+	pr := printer.FromContextOrDie(r.Ctx)
+	kf, err := pkg.ReadKptfile(filesys.FileSystemOrOnDisk{}, r.RunFns.Path)
+	if err != nil {
+		pr.Printf("function not added: Kptfile not exists\n")
+		return
+	}
+	if kf.Pipeline == nil {
+		kf.Pipeline = &kptfile.Pipeline{}
+	}
+	switch r.FnType {
+	case "mutator":
+		if kf.Pipeline.Mutators == nil {
+			kf.Pipeline.Mutators = []kptfile.Function{}
+		}
+		newFns := r.updateFnList(kf.Pipeline.Mutators)
+		kf.Pipeline.Mutators = newFns
```

Comment:
append works on nil lists to all this can be simplified as:

kf.Pipeline.Mutators = r.updateFnList(kf.Pipeline.Mutators)